src/python/pants/engine/rules.py

```diff
@@ -350,6 +279,50 @@ def _uncacheable_rule(*args, **kwargs) -> Callable:
     return inner_rule(*args, **kwargs, rule_type=RuleType.uncacheable_rule, cacheable=False)
 
 
+def rule_helper(func: Callable) -> Callable:
+    """Decorator which marks a static function as a "rule helper".
+
+    Functions marked as rule helpers are allowed to be called by rules and other rule helpers
+    and can `await Get/MultiGet`. The rule parser adds these functions' awaitables to the rule's
+    awaitables.
+
+    There are a few restrictions:
+        1. Rule helpers must be "private". I.e. start with an underscore
+        2. Rule helpers can't be rules
+        3. Rule helpers must be accessed by attributes chained from a module variable (see below)
+
+    To explain restriction 3, consider the following:
+    ```
+        from some_mod import helper_function, attribute
+
+        ...
+
+        some_instance = AClass()
+
+        @rule
+        async def my_rule(arg: RequestType) -> ReturnType
+            helper_function()  # OK
+            attribute.helper()  # OK (assuming `helper` is a @rule_helper)
+            attribute.otherattr.helper()  # OK (assuming `helper` is a @rule_helper)
+            some_instance.helper()  # OK (assuming `helper` is a @rule_helper)
+
+            AClass().helper()  # Not OK, won't collect awaitables from `helper`
+
+            func_var = AClass()
+            func_var.helper()  # Not OK, won't collect awaitables from `helper`
+            arg.helper()  # Not OK, won't collect awaitables from `helper`
```

Comment:
Do the relevant calls need to be `await`'ed in order to be detected? Does it ease parsing at all to require the `await`? Since the helpers don't have their arguments injected, the only reason that they need to be detected would be if they were async.

Reply:
I suppose I don't understand the characteristics of `await`ing the helper very well. 

The intent is to basically move code out of the rules to help with readability and re-use. In that regard, I would almost want to assert that the helpers are synchronous. OTOH making them `async` might yield better performance characteristics.

I think initially we should take a stand on this. You know more about `async` than I do, what's your gut feel?

---

> Do the relevant calls need to be await'ed in order to be detected? 

As of now, they can be either `async` or not

> Does it ease parsing at all to require the await?

I don't believe so, not really

Reply:
Discussed offline, but this bit is critical:
> Since the helpers don't have their arguments injected, the only reason that they need to be detected would be if they were `async`.

99% of the time, helpers will be async in order to actually be useful (otherwise they would have signatures like `-> List[Get[X, Y]]` to return some gets which the caller would actually `await`... and that's awkward probably).